DM-8717: Single frame processing task topic prototypes

mssgill opened the pull request:
Have now cleaned up, 'compactified', and properly inserted relevant links and monospacing where we will later link out for all the single-frame proc code pages (4 of them) so i believe they are ready to be reviewed.

Commit:
Getting options and data refs described properly.

Commit:
Each made more concise and with all parts more contentful.

Commit:
Clarified output data types from procCcd that the butler has access to, and the CL args available to it, same as for all CLT's.

Commit:
Cleaning all these up into a compact form to prepare for next PR.

mssgill commented on calibimg.rst:
```diff
@@ -3,7 +3,9 @@
 
-Given a properly characterized exposure, detect sources, measure their
+Given a properly characterized exposure (which means one with a PSF
+determined, commonly done previous to funning this Task by
```
Oops -- it may be fun, but of course, i meant 'running' here.

mssgill commented on calibimg.rst:
```diff
@@ -51,28 +53,28 @@ Flags and utility variables
 --------
 
-- 	``refObjLoader`` - target=LoadAstrometryNetObjectsTask -   reference object loader
+- 	``refObjLoader`` - default target=LoadAstrometryNetObjectsTask -   reference object loader
```
I used the words 'default target' for all subtasks, to make it clear that it's the default, and not just a variable, but a Task its pointing at, thought this would make it clearer..?

mssgill commented on calibimg.rst:
```diff
@@ -84,11 +86,13 @@ Entrypoint
 Butler Outputs
 ==============
 
+Source catalog of type ``src``.
```
Brief here.

mssgill commented on charimg.rst:
```diff
@@ -18,25 +18,6 @@ Some of its primary functions are to:
 
   - Measure the PSF
```
Note removal of the long sequential description, here.

mssgill commented on isrtask.rst:
```diff
@@ -134,14 +133,16 @@ Butler Inputs
-chip-sized mosaic of all amps including overscans and other
-non-science pixels.
+The inputs to the entrypoint method are the exposure to be corrected
+(in ``raw`` format) and the calibration data products. The raw input
```
Specified the specific technical name for the datatype, here.

mssgill commented on isrtask.rst:
```diff
@@ -134,14 +133,16 @@ Butler Inputs
 Butler Outputs
 ==============
 
+Exposure of type ``postISRCCD``.
```
Also the datatype name.

mssgill commented on isrtask.rst:
```diff
@@ -157,173 +158,12 @@ The `write` flag tells the code to write the post-ISR image file to disk.  In th
+In slightly more detail, what this example does is after setting up
+the flag and utility variable configuration, the code makes several
+calibration exposures that will be used to create the final corrected
+output exposure.  Finally, the output is produced by using the ``run``
```
I am assuming this will link later to the actual 'run' function.

mssgill commented on isrtask.rst:
```diff
@@ -157,173 +158,12 @@ The `write` flag tells the code to write the post-ISR image file to disk.  In th
-makes several calibration exposures that will be used to create the final corrected output exposure.  Finally, the output is produced by using the ``run`` function, inputting the raw exposure and the calibration exposures.
-
-
-Details on some of the corrections available in IsrTask
```
Removed this entire section.

mssgill commented on processccd.rst:
```diff
@@ -4,56 +4,59 @@
-by typing processCcd.py at a shell prompt) executes the steps of how
-an image is processed from raw data to a science-grade images and catalogs that can
-be used in analyses.
+ProcessCcdTask (available as the ``processCcd.py`` ``command line
```
I split this into 2 sentences now -- felt it was unwieldy with so many parentheticals.  Input welcome.

mssgill commented on processccd.rst:
```diff
@@ -69,32 +72,36 @@ Entrypoint
 Examples
 ========
 
+The ``obs_test`` package  models a simple camera with one CCD and includes a data repository containing a few raw images (simulating three visits, two with with the `g`-band filter, and one with the `r`-band one), and some associated calibration data. Its camera consists of a single CCD whose geometry matches a subregion of a single LSST CCD.
```
Have added a bit of info about what's actually in obs_test, though we link to it and they can go there for details as needed.

mssgill commented on processccd.rst:
```diff
@@ -69,32 +72,36 @@ Entrypoint
   processCcd.py $OBS_TEST_DIR/data/input --output processCcdOut --id
 
-The data is read from the small repository in the ``obs_test`` package and written to: ``./processCcdOut`` (or whatever output you specified). Specifying ``--id`` with no values processes all data. Add the option ``--help`` to see more options.
+The data is read from the small repository in the ``obs_test`` package and output images and catalogs are written to subdirectories in: ``./processCcdOut`` (or whatever output name you specified).
```
Very minimal info on the outputs.

mssgill commented on processccd.rst:
```diff
@@ -69,32 +72,36 @@ Entrypoint
+Command Line Arguments 
+======================
 
+ProcessCcdTask has all command line arguments available to a general
```
(I believe the modification to --id is not used because IsrTask hardcodes the datasetType it takes in to 'raw' as: ccdExposure = sensorRef.get('raw')  currently)

Commit:
Encapsulates the changes we all agreed on from the morn telecon.  Will give it one more once-over, then submit a PR and ask for it to be reviewed.

Commit:
Final small changes from suggestions in morn.

mssgill commented on CLTargs.rst:
```diff
@@ -0,0 +1,56 @@
+positional arguments:
```
This is just a 'placeholder' file for now, where i dumped out all the stuff from doing processCcd.py --help (i.e. the standard CLT args) into this file, but it will be cleaned up/replaced by a more correct file later.

mssgill commented on calibimg.rst:
```diff
@@ -55,26 +55,26 @@ Parameters
 
-Subtasks
---------
+Subtask Targets
```
Changed name to 'Subtask Targets' for these sections in all the files, and moved above the 'Parameter' sections.

mssgill commented on calibimg.rst:
```diff
@@ -55,26 +55,26 @@ Parameters
+----------------
 
-- 	``refObjLoader`` - default target=LoadAstrometryNetObjectsTask -   reference object loader
+- 	``refObjLoader`` - default=LoadAstrometryNetObjectsTask -   reference object loader
```
Removed the extra and now unneeded word 'target' from the defaults for all of these subtask target descriptions.

mssgill commented on calibimg.rst:
```diff
@@ -100,6 +100,7 @@ The example code is ``calibrateTask.py`` in the ``$PIPE_TASKS/examples`` directo
 
      python examples/calibrateTask.py --display
      
+Running this example currently requires that over and above the DM Stack installation, ``afwdata`` is installed and set up (via the EUPS ``setup`` command).
```
Added this comment because it's needed for the example to work.

mssgill commented on isrtask.rst:
```diff
@@ -134,14 +135,14 @@ Butler Inputs
 
 The inputs to the entrypoint method are the exposure to be corrected
-(in ``raw`` format) and the calibration data products. The raw input
+(which will be of ``datasetType`` ``raw``) and the calibration data products. The raw input
```
Added that the 'raw' is a specific datasetType.

mssgill commented on isrtask.rst:
```diff
@@ -134,14 +135,14 @@ Butler Inputs
 ==============
 
-Exposure of type ``postISRCCD``.
+Exposure of ``datasetType`` ``postISRCCD``.
```
Also, postISRCCD is a datasetType.

mssgill commented on isrtask.rst:
```diff
@@ -178,6 +179,6 @@ Algorithm details
 -------------
   
-  [Extra reference: Section 4 of LSST DATA CHALLENGE HANDBOOK (2011), and http://hsca.ipmu.jp/public/index.html ]
+  [Extra reference: Section 4 of LSST DATA CHALLENGE HANDBOOK (2011) [https://project.lsst.org/sciencewiki/images/DC_Handbook_v1.1.pdf] , and http://hsca.ipmu.jp/public/index.html ]
```
This link is actually where i got it, and is still live, so in some way, we may want to include this ref, or the appropriate DMTN later.

mssgill commented on processccd.rst:
```diff
@@ -18,11 +18,11 @@ catalogs (``background`` and ``sourceCat`` fields of a
 
 
-- 1.  ``Instrument Signature Removal`` -- Implemented by the :doc:`IsrTask <isrtask>` sub-task (isr configuration), this step removes CCD signatures (such as bias, dark current, flat-fielding, and cross-talk) and masks bad pixels.
+1.  ``Instrument Signature Removal`` -- Implemented by the :doc:`IsrTask <isrtask>` sub-task (isr configuration), this step removes CCD signatures (such as bias, dark current, flat-fielding, and cross-talk) and masks bad pixels.
```
Took out the unordered list directives (the '--') and just left the ordered list numbering.

mssgill commented on processccd.rst:
```diff
@@ -104,4 +105,4 @@ Command Line Arguments
 
 ProcessCcdTask has all command line arguments available to a general
-``command line task``.
+``command line task``, which can be found on :doc:`this page <CLTargs>`.
```
This is link to that placeholder file -- works, though will need replacing in the future.

mssgill commented on calibimg.rst:
```diff
@@ -4,9 +4,10 @@ CalibrateTask
-determined, commonly done previous to funning this Task by
-:doc:`CharacterizeImage <charimg>`), detect sources, measure their
-positions, and do a photometric measurement on them.
+determined and shipped along with the image, commonly done previous to
```
I used the word 'shipped' vs. 'attached', seemed more accurate, but open to input on this.

mssgill commented on calibimg.rst:
```diff
@@ -17,14 +18,35 @@ This task is implemented in the ``lsst.pipe.tasks`` module.
--``doWriteHeavyFootprintsInSources`` (`bool`) - defaults to `True` -
-    Include HeavyFootprint data in source table? If false then heavy
-    footprints are saved as normal footprints, which saves some space
+-``doWriteHeavyFootprintsInSources`` (`bool`) - defaults to `True` - Include HeavyFootprint data in source table? If false then heavy footprints are saved as normal footprints, which saves some space
```
Just correcting the rst so it's rendered properly.

mssgill commented on isrtask.rst:
```diff
@@ -160,9 +160,9 @@ The `ds9` flag tells it to bring up the ds9 image viewer (if installed) and show
-the flag and utility variable configuration, the code makes several
-calibration exposures that will be used to create the final corrected
-output exposure.  Finally, the output is produced by using the ``run``
+the parameter configuration, the code makes several calibration
```
Changed the reference term here as this is what we're now calling these.

Commit:
Making a cleaner subdir with all the files to actually be under review
currently in it.

Commit:
For the rst files so far just:

   "subtask targets" -> "retargetable subtasks"

For conf.py -- copied over one in pipelines_lsst_io, modified it
slightly as so:

- 2016 -> 2017
- commented out release version

Then was getting a warning, found from:
https://github.com/rtfd/readthedocs.org/issues/1776

Modifying this line:

 html_static_path = ['_static']

to:

  html_static_path = []

made it compile fine, no warnings.

Commit:
Get repo into standalone buildable shape

Commit:
Comment currently offending line in requirements.txt so compiling now works.

Commit:
Need this conf.py file from validate_base.

Commit:
Make repo consistent with lsst-sphinx-bootstrap-theme.

Commit:
Make language of subtasks vs. config consistent, and link to each of them.

Commit:
Insert retarg subtask and param tables in csv-table form.

Appeared the best choice between indented list, table (in box form),
and csv-table.  Takes some work to get the formatting correct to get
tables to compile correctly and show up right, but they do appear much
nicer than the previous way the lists were inserted.  Just done for
calibrate so far, other 3 tasks need to get the same workover.

Commit:
Put all retarg subtasks and params into csv-table format.

Have put all these config lists into the new and nicer format for the
4 sfp tasks.  Nested lists in cells not working yet, are a next step,
but want to save this version at least however those work out later.

Commit:
Insert a few words about what the 'background' processCcd returns is.

Commit:
Insert extra comments about task invocation from doxygen.

Because for 3 of the sfp Tasks we already have comments about the
entrypoint/invocation, we should go ahead and include those, we do so
here.

Commit:
Correct backticked items in intro and entrypoint sections.

At least in these sections so far, have put in more proper links, to
either current Doxygen pages, or placeholder ones to pages we will
make, and removing a few, replacing with single backticks for
emphasis.  Should help with properly templating where we want people
to eventually go.

Commit:
Include some level of algorithmic description for the tasks.

This is extremely minimal and probably not really the kind of thing we
want ultimately to guide developers in their docstring-filling, but
it's a very first step at doing this.

Commit:
Reuse doxygen docstrings in our rst file (for procCcd)

This is how we can use the API comments currently in the doxygen
docstrings and incorporate them into our own rst file, using a mockup
written by JSick.  It duplicates other parts of the page (e.g. the
Entrypoint), currently, but has a new section on Initzn, and is a good
first draft of how we might go about this.

(For conf.py: during sphinx processing there was a warning that
version wanted a string not a float, so i changed that to no longer
get the warning.)

Commit:
Make 2 initial API proto pages.

Just done for procCcd, moved JSick's comments into this, and a very
basic version for isr, not complete, but a start.

Commit:
Reorder processCcd so sections accord with DMTN-030

Also with our telecon in the morn, where we decided that the Python
Usage section should come early, so this is our current ordering:

- intro
- config
- python usage
- CL usage [For CLT's]
- butler i/o
- examples
- debugging
- algorithm details

[As an aside: using emacs column number mode now to make sure all
commit msg summary lines will be 50 char or less, and will wrap each
line in body of msg at 72 char]

Commit:
Suppress html_static_path warning during proc.

Did this by inserting html_static_path = [] into the conf.py file.

Commit:
Reorder sections according to DMTN-030

Just shuffling sections to make consistent with our current best idea
of how to implement section ordering.

Commit:
Fill in examples section with code extracts

We want to provide some of the most relevant lines of the example
script code so that users know what's happening when they are running
it, and we so far do this along the lines of the choices the doxygen
string writers did, selecting those same parts.

Also modified links in this section so they point to relevant places,
like the current EUPS guide.

Commit:
Make links in calibimg go to proper destinations

Just straightening out these links so they're all going to the right
places, like in the other 3 sfp Tasks.

Commit:
Change CalibTask to the more proper CalibrateTask

Commit:
Put debug vars properly into a csv-table

So that they are now consistent with the params table.

Commit:
Remove file now incorporated into conf.py

Commit:
Insert code lines from the runIsrTask example

This brings back a fair amount of what i'd written earlier into the
example section to fill it out and detail what's going on with
runIsrTask more thoroughly.  The core is from the doxygen pages, but i
added more detail and interstitial explanation also, which i think is
something useful we can spell out for developers.

Commit:
Move line for afwdata reqt earlier in example sec

Want to start with this right at the start of the sec so that users
will know they need it before trying out the example.  Also linked to
afwdata on github.

Commit:
Mark off all references to pipebase structs

To be consistent with rest of pages, i demarcated off the
``lsst.pipe.base.Struct`` refs, so that we can link these all later to
the standard location.

Linked lsst.pipe.base.CmdLineTask to proper place, also.

Commit:
Put in python usage/initzn section for isrtask

So that we have this taken care of, done in straight analogy with
the same sec for processCcd.

Commit:
Put in python usage/initzn section for charimg

So that we have this taken care of, done in straight analogy with
the same sec for processCcd (and isrtask).

Commit:
Put in python usage/initzn section for calibimg

So that we have this taken care of, done in straight analogy with
the same sec for all 3 of the other Tasks (most similar to charimg).

Commit:
Clean up syntax in charimg

Paren location and list formatting.

Commit:
Remove 2 obsoleted files from dir and ToC

All the content i had in those initial API pages was for users, and
has been incorporated into the main rst pages for those Tasks, so the
files can now be removed fully from the dir, and the index.rst ToC
file, as well.

Commit:
Add properly formatted API file for charimg

Chose to make separate API file for this Task (the one developers
likely would care most about), and this is the first draft with all
the functions formatted correctly.

Commit:
Give a return field to getSchemaCatalogs()

Short function, but to make consistent with the others.

Commit:
Reduce double backticked items + insert links

For these 2 Tasks so far, have moved most of the double backticked
items into single backticked ones, and then inserted relevant links to
doxygen and github pages wherever relevant and easily findable.  A few
links are placeholders, not going anywhere currently, but planned to
go to a page we make in the future, or probably there already exists
one that hasn't been easily found yet.

Commit:
Change double to single backticks for charimg

Now for charimg: have moved most of the double backticked items into
single backticked ones, and again inserted relevant links to doxygen
and github pages wherever relevant and easily findable.  A few links
are placeholders, not going anywhere currently, but planned to go to a
page we make in the future, or probably there already exists one that
hasn't been easily found yet.

Commit:
Change double to single backticks for calibimg

Now for calibimg: have moved most of the double backticked items into
single backticked ones, and again inserted relevant links -- in this
case most links are placeholders, not going anywhere currently,
awaiting consensus that the places the charimg ones go is appropriate
(most calibimg ones are related or the same).

Commit:
Remove parens around var types in config table

Not needed, and to make consistent with other Tasks.

Commit:
Insert cmd line usage section for procCcd

Put in text on how to use from the cmd line, and also all the flags
that one might use there.  Put these in a csv-table format, which took
some hammering them into shape.

Commit:
Tweak all sections to accord with Slack comments.

Some final edits to butler, debug, entrypoint, API usage sections esp,
to be consistent with the comments written up in the Slack document
for this 'release'.

Commit:
Remove entrypoint for isr, insert link elsewhere

Specifically, inserted into the Python Usage section for the run function descrip

Commit:
Make calibrate method target consistent (calibimg)

Just to the doxygen page for it, currently.

Commit:
Make double to single backticks (api_charimg)

This one had escaped the wave of modifying single to double earlier,
now corrected (also small adjusting of indentation for 'bkgd')

Commit:
Correct CLT link

One reference for this had been in error earlier.

Commit:
Integrate 'API Usage' sec into See Also box

Just simply doing this for all 3 non-CL Tasks, no need for a separate
section that only has one link in it.

Commit:
Put general CLT options in a separate file

None of these are specific to procCcd, so we put them in a general CLT
options table file (added here), then refer to them from procCcd.

Commit:
Move CL usage section higher in procCcd

In accord with consensus on ordering from last mtg.  No butler sec for
procCcd (and we're keeping examples above debugging for now).

Commit:
Add :orphan: true directive at top

This prevents Sphinx from complaining in a warning that these
particular files aren't included in a ToCtree (i.e. in index.rst)

Commit:
Remove apiUsage_charimg in index.rst, not needed

Commit:
Change ordering to move debug before examples

This was the group consensus on section ordering.

Commit:
Function -> Method

Commit:
Insert rst comments on links

This is a wholesale entering of multiple rst comments in all sfp files
explaining that we'll want to link to pages for many of the technical
catalog/exposure etc. types.  These aren't given in much detail or
precision other than to say we need them for now, we certainly want to
figure out the framework for these pages.

In other places, comments are given to why the particular doxygen page
was chosen to be linked to.

(In some cases the comments are a few lines down from the ref because
Sphinx gives warnings about lines not being ended properly if i
attempt to interleave them in a list.)

Commit:
Include inputs after run, remove butler sections

Found that the butler sections weren't describing anything more useful
than was in the python usage section, so simply inserted a quick
sentence after the 'run' method interface is described saying what's
required to be input for the Task to work, and referring to the
details below it.

Commit:
Add test versions of 2 stubbed out API linked docs

These are test versions that work for at least the first 2 links after
a fair amount of working over the format; want to keep them for now
while i get the other links right.

Commit:
Back out stubbing attempt for API pages

This worked fine for making a small example, but causes problems with
format and links not working right when inserted in my
already-existing page docs, even after some extensive wrangling with
it, so since it's a nice option but not required for this version,
i've just backed it out.

Commit:
Prevent warning during sphinx compile (char.rst)

Just inserting the orphan: true directive into this file that i was
using during the API stub effort.

Commit:
Revert to original for apiUsage_charimg

Commit:
Fill in algorithmic sections

Have done a fair bit in all of these by tracing through code,
inserting links, and bringing back some older text i had written for
Isr functions.  Will give more details in the Slack Release Note doc.

Commit:
Fill in what will be installed from reqts file

Just to keep the Readme updated.

Commit:
Insert Confluence page with i/o's into isrtask

Inserted Hsin-Fang's page which has useful very concise info into this
specific page, because isrtask has so many inputs, though it could be
somewhat useful for all the sfp pages.  Not sure we should be linking
to a page like this, but in the interest of more info being
potentially useful, i did so for now.

Commit:
Add ref to Coulton et al. report to b-f correction

This is currently an internal report and should be made public, and i
put that into an internal rst doc comment.

Commit:
Add links in for charimg algorithm sec

Cleaned up this sec a bit, and added links for bkgd and cosmics
determination.

Commit:
Round out algorithm descrip for calibimg

Did this by making it more standalone - still refers to charimg, but
now explains what calib does on its own, too.  Further added in the
optional modules that are run (apcorr and deblend) and made links to
them.

Commit:
Add in bit on overscan correc to isrtask example

This was actually on the doxygen page but i'd left it out earlier in
my detailed isr example, want to put it back in so the example is
given in substantial detail, and can be cut back later if desired.

Commit:
Insert needed linebreak for comment string

Otherwise it was showing up in the rendered page, so just had to fix
this.

Commit:
Insert line to conf.py to suppress build warnings

These are the warnings for the src file because the bootstrap-theme is
being pulled into it when pip install -r requirements is run (see
Slack dm-docs discussion of 2-7-2017).

Commit:
Modify link destinations in procCcd

Following KSK recommendations, we now link CLT out to how one
implements them vs. the base CLT class.  We also inserted link to
github src code vs. doxygen version.

Commit:
Move entrypoint section content into run method

Moved all former entrypoint content (mostly about persisting, and the
link to 'run') into the Python Usage section under 'run', as per KSK
suggestions.

Commit:
Insert placeholder Butler link

Will want people to be able to go to some explanation of the Butler,
so just put a place for where actual link will go later.  (Did this in
all 4 Tasks, but did more in charimg and calib, so committed them
separately.)

Commit:
Fix links and insert API link stub for procCcd

Following consensus ideas from the DM Docs Team Thurs mtg,
implemented:

- Removed current link to github src for procCcd.py (we'll insert
  later when we have the API pages properly set up for all tasks)

- Want 'options' in run signature to link to CLT page, but in current
  tech, we can't do this, so for now, commented in rst that we will
  want this

- API page for procCcd: inserted link to where it will go in the See
  Also box, in parallel to the other sfp tasks (even if it turns out
  it will be very minimal, with just the 'run' function given ultimately)

Commit:
Write ``struct`` to prevent linking to python obj

We double backtick ``struct``so it doesn't get directly linked to the
the python obj through the 'default role' in rst, which is not what we
want.

Commit:
Insert double backticks for param obj values

We do this so they are interpreted as code literals (i.e. monospaced)
vs. being displayed in the default role format in rst.  Done for all 3
sfp subtask pages (i.e. not procCcd, which has no such params).

Commit:
Insert comment for plan to link to Framework pages

We later want to link higher level pages in the Framework docs about
CLT's at this location, so just put a placeholder comment in for now.

Commit:
Insert placeholders to API pages for '..Task' refs

Everywhere where the task is mentioned in its own page, we insert the
link to the API page for it (placeholders for isr and calib, and the
draft one for charimg).

jonathansick committed:
Implement a config section as a raw html table

Using raw html tables (that emulate docutils table class structures)
allows us to properly mark up lists and other complex cell content.

Commit:
Make small tweaks to param and subtask tables

Just cleaning up these tables with small changes (removing a quote,
inserting a link, putting in an emphasis)

Commit:
Insert raw html table for params for isrtask

This allows us to make the lists in the subcells look ok, compared to
before (see discussion with JSick in Slack thread in dm-docs on
2/12/2017).

Commit:
Remove and insert double quotes

For consistency with the other str entries in the table, i realized i
needed to make these changes.

Commit:
Link and bold the str in the raw html table

Just to make a step more consistent with the rst part of the table, we
put in the link and insert the html bold tag by hand.

Commit:
Merge branch 'u/jonathansick/config-table' into tickets/DM-8717

SimonKrughoff commented on processccd.rst:
```diff
@@ -4,62 +4,66 @@
+ProcessCcdTask (available as the ``processCcd.py`` ``command line
+task``) executes the steps of how an image is processed from raw
+uncorrected CCD-level data finally to science-grade images and
+catalogs.
```
I think we should always be as precise as we can be.  The term science-grade doesn't really capture this.  Also, I think the word raw and uncorrected are synonyms in this instance.  What about something more like: "executes the processing steps to turn raw pixel data into characterized images and calibrated catalogs"?

mssgill replied:
Ok good wording suggestion, changed.

[I thought to use "pixel-level" vs. just 'pixel' to implicitly indicate that it's also been through the electronics after]  (?)

SimonKrughoff commented on processccd.rst:
```diff
@@ -4,62 +4,66 @@
 
-- The second step is to do ``Image Characterization`` by running :doc:`characterizeImageTask <charimg>`), which includes for our purposes: object detection, repairing of cosmic ray defects, measuring and subtracting of sky background, and then finally measuring bright sources and using this to estimate background and PSF of an exposure.
+
+1.  ``Instrument Signature Removal`` -- Implemented by the :doc:`IsrTask <isrtask>` sub-task (isr configuration), this step removes CCD signatures (such as bias, dark current, flat-fielding, and cross-talk) and masks bad pixels.
```
Again the term "CCD signatures" is not very precise.  It is really all the effects imprinted by the camera.  This includes defects in the CCDs through to electronics.  It's a little confusing because some of the telescope signatures will be removed in this step (vignetting) but not others (subtle differences in the lens coatings).

This paragraph also seems to conflate the effects being removed with the algorithm for removing them.  I think simple is probably better.  Something like:
"The IsrTask subtask applies pixel level corrections in preparation for image characterization.  Corrections include: bias, dark current, cross-talk, flat-fielding and bad pixel masking."

mssgill replied:
Ok - agree that calling it CCD-specific was not exactly correct, certainly from the perspective of someone dealing with the steps of processing raw data until it gets to output imgs, so i've used mostly your wording, but kept the link in to IsrTask in parallel with the other 2 steps.

SimonKrughoff commented on processccd.rst:
```diff
@@ -4,62 +4,66 @@
+
+1.  ``Instrument Signature Removal`` -- Implemented by the :doc:`IsrTask <isrtask>` sub-task (isr configuration), this step removes CCD signatures (such as bias, dark current, flat-fielding, and cross-talk) and masks bad pixels.
+
+2. ``Image Characterization`` -- Implemented by the :doc:`CharacterizeImageTask <charimg>` sub-task (charImage configuration), this step repairs cosmic ray defects, estimates and subtracts a background, does object detection, and estimates a PSF.
```
I don't think of cosmic rays as defects.  Defect used in this context means specifically problems with the physical detector or the electronics attached to it.  Cosmic rays are actual detections.  I would prefer "interpolation of cosmic rays".

The fact that object detection is done in this step is sort of a side effect of the algorithm.  I think it is probably better to make that clear "estimates and subtracts a background and estimates the PSF using a set of high signal-to-noise detections.  This process is iterative."

mssgill replied:
Hm - ok, i've changed the text to "this step interpolates over cosmic rays" and i understand not wanting to call CR's defects, but is "interpolation of cosmic rays" fully clear enough to the reader?  Maybe i need to understand exactly what's done better, but it's something like that they are being masked out in this step by their pixels being  interpolated over, yes?  If so, maybe we should say it this way..?  (And initially, i used the word 'repair' because it's what's used for the routine in charImg itself.)

Your second change is fine, except i do wonder if the final sentence is needed at the top here, where we're trying to be brief.  The fact that it's iterative maybe belongs more to the charImg description itself, it's currently in the algorithm section.  Don't feel strongly on this one, though, so i put that in, too.

SimonKrughoff replied:
As with all my comments, I expect you to reword things.  I don't want you to use things verbatim.

I think "repair" gets the notion across too.  In this process, CRs are detected, masked and interpolated.

mssgill replied:
Ok, i've altered wording here to where i think it gets the idea across best, now.

SimonKrughoff commented on processccd.rst:
```diff
@@ -4,62 +4,66 @@
+2. ``Image Characterization`` -- Implemented by the :doc:`CharacterizeImageTask <charimg>` sub-task (charImage configuration), this step repairs cosmic ray defects, estimates and subtracts a background, does object detection, and estimates a PSF.
   
-- The last step is doing ``Image Calibration`` by running :doc:`calibrateTask <calibimg>`, which measures faint sources, does the astrometry by fitting an improved WCS to the image, and figures out the photometric zero-point for the image.
+3. ``Image Calibration``  -- Implemented by the :doc:`CalibrateTask <calibimg>` sub-task (calibrate configuration), this step measures faint sources, fits an astrometric WCS and extracts a photometric zero-point for the image.
```
It doesn't just measure faint sources.  Again I think this is a point of precision.  This detects, deblends and measures sources in the frame to a configurable signal-to-noise threshold.  Even that is a little imprecise because the detections are done in a likelihood image, so only the point sources are truly optimally detected, but that should be in the algorithmic description of the detection algorithm.

mssgill replied:
Ok, i changed this at the top, and inside the alg. section of calib too. but a pointer to details of why this detection routine is best for point srcs could be helpful here, if there's a place it's specifically described (?)  -- reviewing code it wasn't obvious to me, nor did looking through sec 3.1 of LDM-151 (which seems to be a useful document for this) explain this so clearly.

SimonKrughoff commented on singleFrameProcCoreTasks/isrtask.rst:
```diff
@@ -0,0 +1,432 @@
+Instrument Signature Removal (ISR) is a sequence of steps taken to
+'clean' images of various aspects of defects that any system of optics
+and detectors will imprint on an image by default, and is generally
+one of the very first procedures carried out on an exposure.
```
This will be *the* first pixel processing step in the pipeline (there may be some data/metadata reorganization but that's inconsequential).  Defects is used here in a non-standard way.  In essence the goal here is to correct the image from counts to the the photons that entered the surface of the device (this is broken slightly by the fact that illumination corrections are sometimes applied here).  I.e. no optical distortion or other photometric calibration is done in this step.  I think it would be much cleaner to just say that this step attempts to correct by effects imprinted on the data by the physical characteristics of the detector and the electronics of the sytsem (better worded, of course).

mssgill replied:
Ok... cleaned up and reworded the intro section into 2 paragraphs incorporating some of the above.

SimonKrughoff commented on singleFrameProcCoreTasks/isrtask.rst:
```diff
@@ -0,0 +1,432 @@
+defects will be present and need to be removed, and thus the sequence
+of steps taken will vary from image to image.  Generally these
+corrections are done one CCD at a time, but on all the amplifier
+sub-images at once for a CCD.  
```
I think this paragraph can be merged with the one below.  The sequence of steps only varies from camera to camera (not image to image) to first order.  I think this could be collapsed to just one sentence saying that the corrections will vary (including potentially order) and that IsrTask implements a reasonable reference version.

mssgill replied:
Ok, i condensed according to the above, and collapsed paragraphs.. though my personal preference is *not* to do that, and i do believe in the age and era of web reading, small short paragraphs are much more readable.  (I believe this is espoused by the EPPO philosophy, too..)

SimonKrughoff commented on singleFrameProcCoreTasks/isrtask.rst:
```diff
@@ -0,0 +1,432 @@
+
+`IsrTask <#>`_ provides a generic vanilla implementation of doing these
+corrections, including the ability to turn certain corrections off if
+they are not needed.
```
Words are precious.  I think "generic vanilla" is needlessly redundant.  Similarly, I think you can probably drop "if they are not needed."  The ability to turn them off is the important part.  The user can decide if they are needed or not.

One thing that is not included is that through this process the amp images are trimmed and assembled into a pixel grid ready for characterization and calibration.

mssgill replied:
Ok, trimmed down.

And yes, right, isrtask also calls assembleCcd where the full 16 amps (for LSST) are assembled into a single img for the char and calib steps.  I've included that now.

SimonKrughoff commented on singleFrameProcCoreTasks/isrtask.rst:
```diff
@@ -0,0 +1,432 @@
+
+Algorithm details
+====================
+
```
I think this content as it exists should probably be removed because it requires a lot of work to be the level I think we want.  First off, the descriptions are in words and should be in math.  Second, I don't know where we want to put the narrative description of the effects, maybe it goes here, but I don't think the way it's implemented here is quite right.  Third, the level is very different from section to section here.

Granted this is probably one of the longest algorithmic sections.  I think that actually means we need to put each of these corrections in their own tasks and document the algorithm in the subtask.

mssgill replied:
Ok - all removed, and one bit, for B-F correction, edited and left in.

SimonKrughoff commented on singleFrameProcCoreTasks/isrtask.rst:
```diff
@@ -0,0 +1,432 @@
+(The default DMS method in particular is described in substantial
+detail `here`_, and also even further in a currently internal report
+by Coulton, Lupton, Smith and Spergel from 4-14-2015 and in references
+listed therein.)
```
I think the text that is in the link is what should be in here.  There is still the question of the narrative description of the effect, but I'm not sure that should be here anyway.

mssgill replied:
Incorporated in, now.

SimonKrughoff commented on singleFrameProcCoreTasks/charimg.rst:
```diff
@@ -0,0 +1,304 @@
+
+Given an exposure with defects repaired (masked and interpolated over,
+e.g. as output by :doc:`IsrTask <isrtask>`), this task does initial
+source extraction and PSF estimation.
```
I think it's important that the input is fully ISR corrected, not just defect interpolation.  It may be best just to say that, unless you want to call out specifically that the defects are repaired.

mssgill replied:
Ok, done.  (I think i must have kept using the word  'defect' because it was in some of the initial descriptions i saw.)

SimonKrughoff commented on singleFrameProcCoreTasks/charimg.rst:
```diff
@@ -0,0 +1,304 @@
+
+  - Measure and subtract background
+
+  - Measure the PSF
```
This may be slightly too simplistic.  The process is iterative so that the input is an image with a guess at a PSF and it iterates over the steps to refine the PSF and background model (updating what it thinks is a CR as it goes).

mssgill replied:
Yes, i know that routine in the code, and actually, this is described in the algorithm section below.. but i still don't know that it needs to go at the very top, where we're trying to be brief..?

SimonKrughoff replied:
Again, it's about precision.  We need to be brief but correct.  I think it's fundamental to the algorithm that it is iterative.  I.e. you can't measure without a PSF, but you need detections to measure the PSF.

mssgill replied:
Ok, yep, clearer -- incorporated that now.

SimonKrughoff commented on singleFrameProcCoreTasks/charimg.rst:
```diff
@@ -0,0 +1,304 @@
+
+Algorithm details
+=================
+
```
I think we talked about this, but I'm against calling out specific code implementations here.  I think it should be just math (and the narrative part that I don't know how to deal with).

mssgill replied:
Hm, little unclear on this -- though i link to code if they want to go look up specifics, my description of all the things that are done is pretty general i thought.  

But yes, clearly I haven't inserted any math yet anywhere (and not certain where I would do this anywhere except for say the B-F correction where we do have details..)

jdswinbank replied:
Clearly, as a group, we should discuss what we're hoping to get out of this section in more detail.

Somewhere in the task documentation, I think we need a brief functional description of what's happening. It can assume the reader knows the material in the introduction, and that they understand the ultimate aim of the process (we do not need to educate them about why they want a good background model, for example). I'm thinking of something like the following, perhaps:

> The PSF is iteratively arrived at by repeatedly interpolating over cosmic rays (delegated to the `repair` retargetable subtask),  estimating and subtracting the background (the `background` subtask), detecting (`detection`), deblending (`deblend`) and measuring (`measurement`) sources, and using those sources to estimate the PSF (`measurePsf`). This is repeated `psfIterations` times, gradually refining the PSF model. After the ultimate PSF has been so derived, it is used in final `repair` and `measurement` steps which produce the source catalogue returned to the caller.

(NB scanning the code I'm not sure the above is strictly accurate, but I think it gives the gist of what's appropriate.)

I'm actually not sure whether it's necessary to refer to specific code objects or not. I think it's probably useful, since the reader likely cares that all these algorithms are retargetable, but you could likely persuade me either way.

I think a brief paragraph like the above feels at home in “Algorithmic Details”, but you could probably convince me that it belongs in the introduction or something instead.

In addition, depending on the contents of the task, it _may_ be appropriate to provide one or both of:

- References to published papers, technical notes, etc, describing the algorithm.
- Where external descriptions of the mathematics do not exist, a brief mathematical description of what's happening.

I don't think these are always necessary, particularly for high-level tasks which are effectively aggregations of others (like `CharacterizeImageTask`). Where they are necessary, they form the core of the “details” section.

mssgill replied:
Cool, have incorporated this suggestion into my current version of the charimg algorithm section.

SimonKrughoff commented on singleFrameProcCoreTasks/calibimg.rst:
```diff
@@ -0,0 +1,216 @@
+
+Given a properly characterized exposure (which means one with a PSF
+determined and shipped along with the image, commonly done previous to
+running this Task by :doc:`CharacterizeImage <charimg>`), detect
```
Again, I think this has too much redundancy in the words.  Isn't it sufficient to say "Given a background subtracted image, PSF, and initial astrometric solution, ..."?  You might add a parenthetical link to `CharacterizeImageTask` too.

mssgill replied:
Okay used this wording (though I will say that in this case I think I personally didn't find that sentence to be overly verbose, myself..)

SimonKrughoff commented on singleFrameProcCoreTasks/calibimg.rst:
```diff
@@ -0,0 +1,216 @@
+determined and shipped along with the image, commonly done previous to
+running this Task by :doc:`CharacterizeImage <charimg>`), detect
+sources, measure their positions, and do a photometric measurement on
+them.
```
It's true that measurement is part of this task, but that's different from the astrometric and photometric *calibration*.  That is to say that a major part of this task is to fit models to the object positions to reduce the residual with a reference catalog.

mssgill replied:
Ok.. should this really then be included at the top, in a separate sentence..?

SimonKrughoff commented on singleFrameProcCoreTasks/calibimg.rst:
```diff
@@ -0,0 +1,216 @@
+   
+Algorithm details
+==================
+
```
Same comment as `CharImageTask`.

mssgill replied:
Ok.. altered this section in parallel, also.

Commit:
Insert changes to all sfp codes from KSK comments

These are all changes arising from Simon's review of 2/20/2017, see
the responses at
https://github.com/lsst-dm/scipi_docs_prototype/pull/3 for details.

Commit:
Make sure all default params are code literals

This is now done in all 4 sfp codes, so that everything should render
as a code literal, *except* for 'no default', which i've simply left
as is for now (this is only in isrTask).  I also changed these vars in
the body of the table to code literals where they appear.

Commit:
Remove currently unused files

These were needed when i was experimenting with stub pages, but just
clutter now, thus removed.

Commit:
Make better alg sections for char & calib

These are now more condensed and precise, delegating details to the
subtask links.

Commit:
Remove accidental line left in

(Just needed cleanup)

Commit:
Remove links that refer back to page

These looks broken unless they go to a stub page, so i've removed all
of them, some we'll leave as is, others we'll make simple prototype
pgs for, initially.

Commit:
Redirect links to good explanatory locations

Terms i have found good destination explanatory link pages for:

--- procCcd
- Butler explanation (in all 4 sfp files)

--- isrTask
- postISRCCD

--- charimg
- icSrc
- lsst.afw.math.BackgroundList (less sure of if this link is good, occurs here and in calibimg)
- lsst.afw.table.SourceTable

Commit:
Insert B-F correction algorithm into isrtask

This is as an example of the various algorithms for the isrTask
correction methods.

Commit:
Alter intro sections of all tasks

There was a fair amount of editing on all 4 of the sfp tasks in this
commit -- the intros were altered incorporating github discussions, a
few items changed to code literals that were in the body (like
``None``), and some changes made to clarify algorithm sections also.

Commit:
Change def to indef integral

Had probs with mathjax before on this, but now appears to compile just
fine, so changed all the symbols to indef integrals.

Commit:
Add stub pages for procCcd links

This adds a number of new files where i have just put in simple stubs
for readers to see what *kind* of content we want to put in later.
Now nothing in procCcd links directly to doxygen, but only through
these stub pages.

Commit:
Insert placeholder links for isrTask

Reused some from procCcd, but had to link to several new ones too.
Bit more to do on this, then onto char and calib.

Commit:
Add api for isr stub page

Commit:
Insert more stub page links for charimg and calib

Continuing to update, this is a first complete draft, but needs to be
checked that all links go through correctly.

Commit:
Insert further links on stub pages etc.

This is a lot of the final cleanup changes from mtg ideas on
2-23-2017.  Pages should be in decent shape now, have been checked
through that links work, etc.

Also:

- noted assembleCcd for IsrTask is mentioned but optional, in the intro

- Distinguished for calib that it does Meast and calibration

- Inserted links for defaults of retargetable subtasks
